monai/transforms/transform.py

```diff
@@ -43,15 +45,58 @@
 ReturnType = TypeVar("ReturnType")
 
 
+def _log_pending_info(
+    data: Any, transform: Any, activity: str, lazy: bool | None = None, logger_name: str | None = None
+):
+    if logger_name is None:
+        return
+    logger = get_logger(logger_name)
+
+    if isinstance(transform, LazyTrait):
+        if lazy is not None and lazy != transform.lazy:
+            tlazy = f", transform.lazy: {transform.lazy} (overridden)"
+        else:
+            tlazy = f", transform.lazy: {transform.lazy}"
+    else:
+        tlazy = ", transform is not lazy"
```

Comment:
this kind of transform handling logic in `_log_pending_info` is duplicated in `_apply_transform`, it's not easy to read and maintain, let's think about a better way.

Reply:
It is better out of the main logic than wrapped up with it

Reply:
evaluate_with_overrides actually mixes up logging and logic. I think the refactored code is more decoupled than that

Reply:
It'll be difficult to maintain, cause mismatch of the actual behaviour and the logs. That'll be a new issue introduced by this change.

Reply:
I'm not sure what you mean. What mismatch of the actual behaviour and the logs? This code doesn't change any state on anything. It just logs what is happening, so I don't see how the logs and the behaviour will be different, unless there is a bug here that I am not seeing

Reply:
Are you referring to the fact that there are three lazy modes, False (Force Non lazy), True (Force Lazy) and None (do what the transforms themselves say)? If so, that is the intended design

Reply:
for example, this logging says "(overriden)" but the actual behaviour is not in this function and when reading this code, we never know the actual overriding behaviour, and we don't know where to check the logic.

or another example, `_apply_transform` supports recursively checking a nested structure of dictionaries/tuples, but the `_log_pending_info` doesn't have this logic.
so this already creates various inconsistencies. My original implementation doesn't have this issue.

Reply:
overridden means that the lazy mode of the transform is being overridden by the Compose lazy mode

Reply:
The behaviour is not in the logging function because it is a logging function

Reply:
I mean this `tlazy = f", transform.lazy: {transform.lazy} (overridden)"` in `_log_pending_info `

Reply:
I'm not following. This is a substring that gets formatted and then put into the main logging string.

```python
    if isinstance(transform, LazyTrait):
        if lazy is not None and lazy != transform.lazy:
            tlazy = f", transform.lazy: {transform.lazy} (overridden)"
        else:
            tlazy = f", transform.lazy: {transform.lazy}"
    else:
        tlazy = ", transform is not lazy"

    if isinstance(transform, MapTransform):
        for k in transform.keys:
            pcount = len(data[k].pending_operations) if isinstance(data[k], MetaTensor) else 0
            logger.info(
                f"{activity} - lazy mode: {lazy}, key: '{k}', "
                f"pending: {pcount}, upcoming '{transform.__class__.__name__}'{tlazy}"
            )
    else:
        pcount = len(data.pending_operations) if isinstance(data, MetaTensor) else 0
        logger.info(
            f"{activity} - lazy: {lazy}, " f"pending: {pcount}, upcoming '{transform.__class__.__name__}'{tlazy}"
        )
```

If it is a lazy transform, we log whether the transforms.lazy is different from the compose's lazy mode. If it is, we note that the transform's .lazy is being overridden on the call.
That tlazy string then gets inserted into the full logging string.

Reply:
that's just an example that the substring should be agnostic to the actual lazy resampling logic, 

if you don't understand this example, then please think about the other example I provided, it may help you understand the design issue here.

Reply:
This is actually informing the reader that although the transform has 'lazy' True it is being overridden by the pipeline executing with 'lazy' False

Reply:
for example, if we have a list of image patches, the logging is incorrect...
```
import torch

import monai.transforms as mt

d = [{"image": torch.zeros((1, 40, 40, 40))}, {"image": torch.zeros((1, 40, 40, 40))}, {"image": torch.zeros((1, 40, 40, 40))}]

xform = mt.Compose(
    [
        mt.FlipD(keys="image", spatial_axis=0),
        mt.Rotate90D(keys="label", k=1, allow_missing_keys=True),
        mt.ZoomD(keys=("image", "label"), zoom=0.8, allow_missing_keys=True),
        mt.SpacingD(keys="image", pixdim=1.2),
    ],
    lazy=True,
    logger_name="log"
)
xform(d)
```

Reply:
this is still an issue as I run it with the latest commit (db1486bfc7be170f1a6517b3194b84e448e696ba), the expected logging (from the dev branch):

```
2023-05-04 10:01:10,000 - INFO - Lazy - 'image' - upcoming: 'Zoomd'- pending 1
2023-05-04 10:01:10,000 - INFO - Lazy - 'image' - upcoming: 'Zoomd'- pending 1
2023-05-04 10:01:10,000 - INFO - Lazy - 'image' - upcoming: 'Zoomd'- pending 1
2023-05-04 10:01:10,005 - INFO - Lazy - 'image' - upcoming: 'Spacingd'- pending 2
2023-05-04 10:01:10,006 - INFO - Lazy - 'image' - upcoming: 'Spacingd'- pending 2
2023-05-04 10:01:10,006 - INFO - Lazy - 'image' - upcoming: 'Spacingd'- pending 2
2023-05-04 10:01:10,019 - INFO - Evaluated - 'image' - up-to-date for - final output
2023-05-04 10:01:10,027 - INFO - Evaluated - 'image' - up-to-date for - final output
2023-05-04 10:01:10,035 - INFO - Evaluated - 'image' - up-to-date for - final output
```

Reply:
the logging results are incorrect for the use case I provided above. that's also a sign of inflexible API design. now I'm worried about this refactoring, at this speed of development, it can't be finished as promised in the previous meeting.

Reply:
How are they wrong?

Reply:
```
2023-05-04 12:36:35,087 - INFO - Accumulate pending transforms - lazy mode: True, key: 'image', pending: 0, upcoming 'Flipd', transform.lazy: False (overridden)
2023-05-04 12:36:35,090 - INFO - Accumulate pending transforms - lazy mode: True, key: 'image', pending: 0, upcoming 'Flipd', transform.lazy: False (overridden)
2023-05-04 12:36:35,090 - INFO - Accumulate pending transforms - lazy mode: True, key: 'image', pending: 0, upcoming 'Flipd', transform.lazy: False (overridden)
2023-05-04 12:36:35,091 - INFO - Accumulate pending transforms - lazy mode: True, key: 'image', pending: 1, upcoming 'Zoomd', transform.lazy: False (overridden)
2023-05-04 12:36:35,094 - INFO - Accumulate pending transforms - lazy mode: True, key: 'image', pending: 1, upcoming 'Zoomd', transform.lazy: False (overridden)
2023-05-04 12:36:35,096 - INFO - Accumulate pending transforms - lazy mode: True, key: 'image', pending: 1, upcoming 'Zoomd', transform.lazy: False (overridden)
2023-05-04 12:36:35,098 - INFO - Accumulate pending transforms - lazy mode: True, key: 'image', pending: 2, upcoming 'Spacingd', transform.lazy: False (overridden)
2023-05-04 12:36:35,101 - INFO - Accumulate pending transforms - lazy mode: True, key: 'image', pending: 2, upcoming 'Spacingd', transform.lazy: False (overridden)
2023-05-04 12:36:35,103 - INFO - Accumulate pending transforms - lazy mode: True, key: 'image', pending: 2, upcoming 'Spacingd', transform.lazy: False (overridden)
2023-05-04 12:36:35,112 - INFO - Pending transforms applied: key: 'image', applied_operations: 3
2023-05-04 12:36:35,118 - INFO - Pending transforms applied: key: 'image', applied_operations: 3
2023-05-04 12:36:35,123 - INFO - Pending transforms applied: key: 'image', applied_operations: 3
```

Reply:
I'd say these aren't missing the first transform (missing out the first transform is a bug, frankly) and provide additional information. Furthermore, given that logging must occur for two separate execution strategies, repeating the logging functionality across multiple paths would be less than ideal. It isn't inflexible design to have a function that handles logging of an object. Its more inflexible to embed that logging in with the logic as opposed to decoupling it. That's how the first transform ended up being missed from the logging

Reply:
Furthermore, if we now want different levels of verbosity (say instead of logging the number of pending transforms we log a list of the names) we can add that to the logging code. If you put that kind of logic in and around the actual logic that does the work, no-one will be able to make sense of what is actual logic and what is logging

Reply:
the code has been changed after my previous comment, and the logging output is now different... just another example showing the logging mechanism in this PR is error prone...

Reply:
Wenqi. I moved the function to a different file...

Reply:
I think it would be most productive to focus on ensuring that the actual behaviour is correct for the pipelines that require it, to avoid having to re-benchmark